### PR TITLE
♻️ Volumes keep track if the data requires saving 🚨

### PR DESCRIPTION
**Pull request by GitHK**
## What do these changes do?

- ✨ added a `.agent` file to each volume containing information about the state of the data on the volume answering the following questions:
  - does the data require saving?
  - was the data already saved?
- ♻️ in addition to sending the `can_save` parameter when closing a services, it is also sent when opening one. In case of unexpected errors (like docker swarm partitions) and subsequent recovery, this comes handy to figure out what to do with the volumes.

**NOTE:** tried to have something nice, but an issue in docker is blocking me from adding immutable files to volumes so user services are not able to modify them at all. See issue https://github.com/moby/moby/issues/45177

## Related issue/s

- https://github.com/ITISFoundation/osparc-issues/issues/638
- required by https://github.com/ITISFoundation/osparc-simcore/pull/3941
- requires https://github.com/ITISFoundation/osparc-simcore/pull/4055 to stop failing CI
## Checklist

- [X] Unit tests for the changes exist
- [x] Runs in the swarm

### CODE REVIEW

File: packages/models-library/src/models_library/volumes.py
Comment (pcrespov) [inline packages/models-library/src/models_library/volumes.py:4]: STYLE: VolumeID suggest to me some sort of uuid or integer ... What about ``VolumeType`` or ``VolumeKind`` , ``VolumeFlavour`` , ``VolumeCategory``

Comment (pcrespov) [reply packages/models-library/src/models_library/volumes.py:4]: 2nd. If the enum value is just used as unique identifier why not using our wonderful ``StrAutoEnum``?

File: packages/service-library/src/servicelib/volumes_utils.py
Comment (pcrespov) [inline packages/service-library/src/servicelib/volumes_utils.py:16]: This field has three possible states: True, False and None. I do not understand the last ... and the description is half way written

File: packages/service-library/tests/test_volumes_utils.py
Comment (pcrespov) [inline packages/service-library/tests/test_volumes_utils.py:12]: ``tmpdir``!?? seriously .. :-)))

Pls use ``tmp_path: Path``

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/volume_files.py
Comment (pcrespov) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/volume_files.py:50]: I understand what you do but read very very strange having a with statment with ellipsis!

I wonder if we are not overdoing with async call and here we should just use
```python
file.touch( mode=(stat.S_IRUSR | stat.S_IWUSR | stat.S_IRGRP | stat.S_IWGRP | stat.S_IROTH), exist_ok=True)
```
would make the trick here and is pretty clear!!

**Review by pcrespov — COMMENTED** [review]
Some comments ... will foolow up later with more.

Shame: I had a nice refactoring of all this part some time ago and go buried under a pile of PRs ... :-(

File: packages/service-library/src/servicelib/volumes_utils.py
Comment (sanderegg) [reply packages/service-library/src/servicelib/volumes_utils.py:16]: also I am not sure why there are 2 booleans?

File: packages/service-library/tests/test_volumes_utils.py
Comment (sanderegg) [inline packages/service-library/tests/test_volumes_utils.py:20]: what means ```requires_saving=True``` and ```was_saved=True``` ?? how can one interpret this?

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/volume_files.py
Comment (sanderegg) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/volume_files.py:50]: this looks very weird indeed. without your @pcrespov comment.. I was wondering what this is about... and I am pretty sure aiofiles might event have the async version of touch...

File: services/web/server/src/simcore_service_webserver/projects/projects_api.py
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/projects/projects_api.py:235]: in case user_role is None. is this actually possible??? you do not need to call in the DB right?

**Review by sanderegg — COMMENTED** [review]
nice, but I think I would need some direct explanations on what this is doing.

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/volume_files.py
Comment (GitHK) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/volume_files.py]: So the only touch implementation is the one from pathlib.

@pcrespov fyi: last time I avoided used pathlib it took me days to find a subtle bug and I had to revert back to aiofiles. So I will figure out how to make pathlib's touch async.

Comment (GitHK) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/volume_files.py]: OK, using PC's suggestion seems to be working.

File: services/web/server/src/simcore_service_webserver/projects/projects_api.py
Comment (GitHK) [reply services/web/server/src/simcore_service_webserver/projects/projects_api.py]: The user_id always exists, so ti will always have a role here. Good point. I've also redacted this piece of code to make less calls to the db.

File: packages/service-library/src/servicelib/volumes_utils.py
Comment (pcrespov) [inline packages/service-library/src/servicelib/volumes_utils.py:19]: OK I understand it now: ``was_saved`` is complementing  ``requires_saving=True``.   I still find it unintuitive.  IMO the problem is the design of the model itself. It reminds me a discussion in  "Product and Sum Types" (pg. 53) of Viafore's book "Robust Python" ( I think we even already discussed about this in the past). I think you should have a look at it for more insight on how to design settings in a scenario like this.

If you do not want to change the design, I would rather go for **``was_saved=False`` as default value** instead of ``None`` since if saving is NOT required then ``was_saved=False`` also makes sense. And if saving IS required, the initial state of ``was_saved`` I guess still has to be ``False`` until the user changes it.  Nonetheless, even in this case it feels strange because, as I said above, I think the design is faulty.

File: services/agent/src/simcore_service_agent/core/settings.py
Comment (pcrespov) [inline services/agent/src/simcore_service_agent/core/settings.py:36]: Q: what about having a constant for ``key_values.json``?

**Review by pcrespov — APPROVED** [review]
Since we have a critical week regarding releases, I would like to assess how critical any new PR getting to master is.

As far as I can tell, this is a critical change that shall be target-tested in staging before it reachs production. If so, I would kindly ask to:
1. actively test it in master and apply all necessary fixes ASAP
2. mark it with 🚨 so it can be target-tested when released to staging

thx

File: packages/service-library/src/servicelib/volumes_utils.py
Comment (sanderegg) [inline packages/service-library/src/servicelib/volumes_utils.py:15]:
```suggestion
            "when True, the contents of the volume must be saved when "
```

File: packages/service-library/src/servicelib/volumes_utils.py
Comment (sanderegg) [inline packages/service-library/src/servicelib/volumes_utils.py:25]: now reading that description I start understand it.
so requires_saving is the flag that defines whether saving will happen or not. Whatever the value of was_saved is. just wondering would it not make sense to have ```last_saved: Optional[datetime]``` ? it is always only once. then maybe not.

File: packages/service-library/src/servicelib/volumes_utils.py
Comment (sanderegg) [inline packages/service-library/src/servicelib/volumes_utils.py:38]: so for instance I do not understand the usefulness of such a validation. if requires_saving is False, then we do not care about was_saved right? is there any other use case I'm missing here?

File: packages/service-library/src/servicelib/volumes_utils.py
Comment (sanderegg) [inline packages/service-library/src/servicelib/volumes_utils.py:41]: here I agree with Pedro. just make was_saved default to False.
Is the only problem not if you initialize with both to True?

File: services/director-v2/src/simcore_service_director_v2/api/routes/dynamic_services.py
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/api/routes/dynamic_services.py:144]: is it can or must?

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_events_utils.py
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_events_utils.py:139]: here I do not understand that the call allows to change ```requires_saving```.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_abc.py
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_abc.py:82]: must

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_task.py
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_task.py:73]: must

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/api/volumes.py
Comment (sanderegg) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/api/volumes.py:26]: it is not a patch since you actually replace it right?
then it is a PUT call or else you need to change your PATCH pattern to allow for selectively changing stuff.

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/volume_files.py
Comment (sanderegg) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/volume_files.py:21]: you can do the same for calling pathlib.touch

**Review by sanderegg — COMMENTED** [review]
please check my comments before merging.

File: packages/service-library/src/servicelib/volumes_utils.py
Comment (GitHK) [reply packages/service-library/src/servicelib/volumes_utils.py]: I took the time to simplify the design. Let me know if makes more sense now.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_events_utils.py
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_events_utils.py]: let me know if now it is better

File: packages/service-library/src/servicelib/enum_utils.py
Comment (sanderegg) [inline packages/service-library/src/servicelib/enum_utils.py:12]: ah this sucks. @pcrespov can't we prevent this? 
if it remains the only one it's maybe ok... but let's keep it as the very only one.

File: packages/service-library/src/servicelib/file_constants.py
Comment (sanderegg) [inline packages/service-library/src/servicelib/file_constants.py:1]: here I am a bit confused by the module name.
we are here in servicelib. file_constants.py. and these are only related to the simcore_agent files right?
maybe rename that to simcore_agent_constants.py at least?

File: packages/service-library/src/servicelib/volumes_utils.py
Comment (sanderegg) [inline packages/service-library/src/servicelib/volumes_utils.py:6]: same for this module's name.

maybe rename to
simcore_agent_volume_utils.py ? or at least something clearer.

Comment (sanderegg) [reply packages/service-library/src/servicelib/volumes_utils.py:6]: same for this module's name.

maybe rename to
simcore_agent_volume_utils.py ?

This is used by both the agent and sidecar is that correct?

File: services/director-v2/src/simcore_service_director_v2/models/domains/dynamic_services.py
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/models/domains/dynamic_services.py:41]: I see.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_public.py
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_public.py:404]: maybe go a bit down with the logs. they tend to clutter a lot.

**Review by sanderegg — COMMENTED** [review]
please rename the modules in servicelib to something clearer. thanks!

**Review by sanderegg — APPROVED** [review]
please rename the modules in servicelib to something clearer. thanks!

File: packages/service-library/src/servicelib/enum_utils.py
Comment (GitHK) [reply packages/service-library/src/servicelib/enum_utils.py:12]: We have na issue for this https://github.com/ITISFoundation/osparc-simcore/issues/4013

**Review by matusdrobuliak66 — APPROVED** [review]

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_public.py
Comment (GitHK) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/api_client/_public.py:404]: I would do that in a separate PR. Not here.

**Pull request merged**
